Scripts/generate-test-config.sh

```diff
@@ -1,14 +1,269 @@
+    return
+  fi
+
+  log "[D]:: $*"
+}
+
+# Always logged
+function log_error {
+  log "[E]:: $*"
+}
+
+# Always logged, except in quiet mode
+function log_info {
+  if [[ $LOG_LEVEL -lt LOG_LEVEL_INFO ]] ; then
+    return
+  fi
+
+  log "[I]:: $*"
+}
+
+function log_trace {
+  if [[ $LOG_LEVEL -lt LOG_LEVEL_TRACE ]] ; then
+    return
+  fi
+
+  log "[T]:: $*"
+}
+
+function die {
+  log_error $*
   exit 1
+}
+
+# Usage & help
+
+function print_usage {
+  echo "usage: ${SYNOPSIS}"
+}
+
+function print_help {
+  usage=$(print_usage)
+  cat <<EOF
+${usage}
+
+SYNOPSIS
+    ${SYNOPSIS}
+
+$SCRIPT_NAME generates a test configuration file for AWS SDK and
+Amplify integration tests.
+
+REQUIREMENTS
+    $SCRIPT_NAME requires the following utilities to be installed
+    and available on your PATH:
+        - Python - v3.7 or higher, accessible as 'python3'
+          (https://www.python.org/)
+    
+    If you use the -a option to assume the test execution role prior to
+    generating the test configuration file, you must also have the following
+    utilities installed:
+        - jq - A JSON parsing utility
+          (https://stedolan.github.io/jq/)
+        - AWS CLI - for retrieving secret values from STS
+          (https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
+
+OPTIONS:
+    -a
+        Assume the test execution role before generating the test configuration
+        file. Requires your environment to have exported credentials for a role
+        or user that has 'sts:AssumeRole' permissions for the integration test
+        execution role. If not specified, temporary role credentials must be
+        exported into the environment variables 'AWS_ACCESS_KEY_ID',
+        'AWS_SECRET_ACCESS_KEY', and 'AWS_SESSION_TOKEN'.
+
+    -p <android|ios>
+        Generate configuration file for platform="<val>"
+
+    -r <region>
+        Look for resources in <region>. If not present, use the value of the
+        environment variable AWS_DEFAULT_REGION. Either the environment
+        variable or this flag must be specified.
+
+    STANDARD OPTIONS:
+    -h
+        Print this help message
+    -q
+        Quiet mode, only errors will be logged
+    -v
+        Verbose output. Specify up to 2 times for more debugging output
+EOF
+}
+
+
+##################################################
+# PARSE ARGUMENTS
+
+assume_role=""
+platform=""
+region=${AWS_DEFAULT_REGION}
+
+while getopts "ap:qvh" optchar ; do
+  case "$optchar" in
+    a)
+      assume_role=1
+      ;;
+    p)
+      platform=$OPTARG
+      ;;
+    r)
+      region=$OPTARG
+      ;;
+
+    q)
+      LOG_LEVEL=-1
+      ;;
+    v)
+      LOG_LEVEL+=1
+      ;;
+    h)
+      print_help
+      exit 0
+      ;;
+    *)
+      print_usage >&2
+      exit 1
+      ;;
+  esac
+done
+
+
+##################################################
+# VALIDATIONS AND ASSIGNMENTS
+
+cmd_quiet_flag="--quiet"
+if [[ $LOG_LEVEL -gt 1 ]] ; then
```

Comment:
It hardly matters, but as a factoid, `[[` is an external program, while `[` is a Bash built-in. So it is more efficient to use `[`, when you can. `[` does support `-z`, `-gt`, etc., as you use in this block.

Reply:
I never realized `[[` isn't a builtin--the docs certainly imply it's a builtin . I'm going to keep it for now since it supports standard constructs like `[[ $foo -eq "foo" ]]` where $foo isn't actually set, whereas `[ $foo -eq "foo" ]` generates a 'unary operator' error unless I wrap $foo in quotes (`[ "$foo" -eq "foo" ]`), which I always neglect to do.